### PR TITLE
feat: toFieldPath policies, replacing upstream MergeOptions

### PR DESCRIPTION
**Pull request by phisco**
Fixes https://github.com/crossplane-contrib/function-patch-and-transform/issues/90.
Almost identical to the upstream implementation.
Except for the renaming:
```yaml
policy:
  mergeOptions:
    keepMapValues: true
```
is now:
```yaml
policy:
  toFieldPath: Merge
```
While,
```yaml
policy:
  mergeOptions:
    appendSlice: true
```
is now:
```yaml
policy:
  toFieldPath: AppendArray
```
The default being `Replace`.
A follow up PR to the migration tooling in the `crossplane` CLI will be needed.

### CODE REVIEW

**Comment by negz** [comment]
Damn, I was really hoping we could avoid this and let SSA do all the merging etc. Obviously we can't do that in this context though, where we're just merging patches into the base template before even applying anything.

A few thoughts:

* I'd like to fix the naming of these options, since moving to this function is a breaking change. See https://github.com/crossplane/crossplane/issues/2581.
* The README says `mergeOptions` are no longer supported, which won't be true anymore.
* It would be useful to make it clear that these options are only used to merge the patch into the base, not to merge the patch into the existing resource. Maybe just in the Godoc for the `mergeOptions` field?

**Comment by negz** [comment]
> I'd like to fix the naming of these options

Maybe something like:

```yaml
      policy:
        fromFieldPath: Required
        toFieldPath: Merge
```

The options for `toFieldPath` would be:

* `Replace`, the default behavior.
* `AppendArray` - equivalent to `mergeOptions.appendSlice`
* `MergeObject` - equivalent to `mergeOptions.keepMapValues`

I could imagine just `Replace` and `Merge` as options too. If it was an array we'd "merge" the two arrays by appending.

**Comment by phisco** [comment]
I do agree we could improve names, but changing them would mean a lot of confusion for users, we could start implementing them as they are, I think the number one goal for this function is a smooth transition path from native p&t. We could think of deprecating the field upstream and push people to the new `toFieldPath` policy first and handle it here too.

**Comment by negz** [comment]
@phisco Please let's fix these names while we have the chance. We've already made breaking changes in this function (e.g. making fields that weren't required before required), and we have a migration tool we can update to handle the renaming.

**Comment by negz** [comment]
> we should at least properly validate the input against the crd to be sure we don't silently ignore it because a user is using the old name by mistake.

Yeah I was thinking the same. Silently accepting unknown fields makes this worse. A quick fix (easier than validating using the CRD) would be to use https://pkg.go.dev/encoding/json#Decoder.DisallowUnknownFields.

> We should probably think about doing it upstream, but implementing it here will take less time. Wdyt?

You mean addressing https://github.com/crossplane/crossplane/issues/2581? Sure. If an API like the one I proposed works we could just mark `mergeOptions` deprecated and introduce the new `toFieldPath`.

**Comment by phisco** [comment]
> You mean addressing crossplane/crossplane#2581? Sure. If an API like the one I proposed works we could just mark mergeOptions deprecated and introduce the new toFieldPath.

I meant validating against the CRD, but your option of strictly decoding is indeed easier. Regarding the API, yes, I think it looks good to me. Nothing against moving forward with that upstream.

**Comment by phisco** [comment]
opened https://github.com/crossplane/function-sdk-go/pull/121 to allow setting `RejectUnknownMembers` here.

**Comment by MisterMX** [comment]
Would be great if this could be merged. Readding `mergeOptions` would make migration from c/c much easier.

**Comment by MisterMX** [comment]
Note for a follow-up PR: If we stick with `AppenSlice` etc. we might want to add a `PrependSlice` option as well.

**Comment by phisco** [comment]
Ok, I switched to the suggested names, let me know if there is anything else, @negz 🙏

File: README.md [hunk 1]
```diff
@@ -128,11 +128,12 @@ These fields are now required. This makes P&T configuration less ambiguous:
...
 
 Also, the `resources[i].patches[i].policy.mergeOptions` field is no longer
-supported.
-
-Composition functions use Kubernetes server-side apply to intelligently merge
-arrays and objects. This requires merge configuration to be specified at the
-composed resource schema level (i.e. in CRDs) per [#4617].
+supported. The same capability can be achieved by setting
+`resources[i].patches[i].policy.toFieldPath` to:
+- `Merge` - equivalent to
```
Comment (negz) [inline README.md:133]: I think it's `MergeObject`, right?

Comment (phisco) [reply README.md]: yes, fixed

**Review by negz — APPROVED** [review]

**Pull request merged**